src/harvester/client.rs

```diff
@@ -0,0 +1,180 @@
+use std::env::var;
+use std::fmt;
+use std::future::Future;
+use std::sync::Arc;
+
+use anyhow::{Error, Result};
+use async_compression::tokio::{bufread::ZstdDecoder, write::ZstdEncoder};
+use bytes::Bytes;
+use cap_std::fs::Dir;
+use reqwest::Client as HttpClient;
+use tokio::time::{sleep, Duration};
+use tokio::{
+    fs::File as AsyncFile,
+    io::{AsyncReadExt, AsyncWriteExt, BufReader},
+};
+
+#[derive(Clone)]
+pub struct Client {
+    http_client: Option<HttpClient>,
+    dir: Arc<Dir>,
+}
+
+impl Client {
+    pub fn start(dir: &Dir) -> Result<Self> {
+        let http_client = match var("REPLAY_RESPONSES") {
+            Err(_err) => {
+                let http_client = HttpClient::builder()
+                    .user_agent("umwelt.info harvester")
+                    .timeout(Duration::from_secs(300))
+                    .build()?;
+
+                let _ = dir.remove_dir_all("responses");
+                dir.create_dir("responses")?;
+
+                Some(http_client)
+            }
+            Ok(_val) => None,
+        };
+
+        let dir = Arc::new(dir.open_dir("responses")?);
+
+        Ok(Self { dir, http_client })
+    }
+
+    pub async fn make_request<'a, A, F, T, E>(&'a self, key: &str, mut action: A) -> Result<T>
+    where
+        A: FnMut(&'a HttpClient) -> F,
+        F: Future<Output = Result<T, E>>,
+        T: Response,
+        E: Into<Error> + fmt::Display,
+    {
+        match &self.http_client {
+            Some(http_client) => {
+                let response = retry_request(|| action(http_client)).await?;
+
+                let file = self.dir.create(key)?;
+
+                let mut file = ZstdEncoder::new(AsyncFile::from_std(file.into_std()));
+                file.write_all(response.as_ref()).await?;
+                file.shutdown().await?;
+
+                Ok(response)
+            }
+            None => {
+                let file = self.dir.open(key)?;
```

Comment:
Shouldn't we have a fall-back to request the actual response in case "file" doesn't exist? Otherwise you run into problems whenever you change your harvester.toml

Reply:
I thought about this too and decided for no fallback because this is a signal that the set of requests changed, i.e. I expect this to mostly happen when the "request key" changes because e.g. the pagination changed.

I also suspect most changes to `harvester.toml` during production would switch one source off and another one on so that running the harvester once without `REPLAY_RESPONSES=` set seems bearable.

Finally, I would like to err on the side of simplicity (we can always add a fallback later on) for now as it makes the whole system easier to understand, i.e. is all of that data from disk or was it partially downloaded? Did I mess up the pagination? Or accidentally changed the request key? These questions seem easier to answer to me if we keep things simply all-network or all-disk for now.

Reply:
(I think creating an issue for adding a network fallback to the  replay-responses feature would be good to ear mark this for future reconsideration though.)

Reply:
Just to write this down: I also tried a version that did not delete and repopulation the responses folder thereby accumulating all responses until they are manually deleted. I decided against this for similar reasons: It would be unclear whether these response were from different network sessions or a single one and therefore harder to understand things if anything went wrong. Additionally, we would have to have an additional function to periodically clean up the folder on the production servers to avoid runaway storage consumption.

Reply:
See #60 for an implementation and let's continue the discussion there.